src/numerics/petsc_matrix.C

```diff
@@ -1201,6 +1201,53 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::mat_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
+                        const std::vector<numeric_index_type> & rows,
+                        const std::vector<numeric_index_type> & cols,
+                        const T scalar)
+{
+  libmesh_assert_equal_to(spm.m(), rows.size());
+  libmesh_assert_equal_to(spm.n(), cols.size());
+
+  for(unsigned int i = 0; i<rows.size(); ++i)
+  {
+    std::vector<numeric_index_type> indices;
+    std::vector<T> values;
+    spm.get_row(i, indices, values);
+    for (unsigned int j=0; j<indices.size(); ++j)
+      this->add (rows[i], cols[indices[j]], scalar*values[j]);
```

Comment:
I am not sure about where you are using this function, but for me, `spm` seems to be small since you are using `std::vector<numeric_index_type>` to map local indices to global indices.

OK, if `spm` is a parallel matrix, we should not use `std::vector`   since `cols` is sparse. If you do this way, you will run out of memory really soon. 

Hopefully, I overreact.

Reply:
@dewenyushu It might be good if you could explain what you are trying to do, since sparse matrix-matrix multiplication is rather unusual. I had assumed that @fdkong would be familiar with the preconditioner you are trying to implement, but it sounds like he's not. I guess it is somehow related to the Schur complement/condensing out the Lagrange multiplier DOFs, so there may be a more efficient way to go about this.

Reply:
@jwpeterson you are correct, both functions are related to condensing out the LM dofs. In order to do so, I need to extract part of the original Jacobian matrix (e.g., D and M shown below), do some calculations (e.g., MDinv*K1ci, etc), and put the new block to the corresponding indices of the condensed Jacobian (shown on the right). 

![Screen Shot 2020-07-08 at 1 19 15 PM](https://user-images.githubusercontent.com/58825741/86960979-a4b8ac80-c11d-11ea-989c-10ba20962ea8.png)

Reply:
Oh, my fault, I ignored the context here. Yes, then `spm` is a regular parallel sparse matrix.

`GetRow, and MatSetValues` should work fine. 

But you have to change to use 


```
const std::map<numeric_index_type> & rows,
const std::map<numeric_index_type> & cols,
```

Reply:
> I am not sure about where you are using this function, but for me, `spm` seems to be small since you are using `std::vector<numeric_index_type>` to map local indices to global indices.
> 
> OK, if `spm` is a parallel matrix, we should not use `std::vector` since `cols` is sparse. If you do this way, you will run out of memory really soon.
> 
> Hopefully, I overreact.

 The row number of spm is the number of dofs on the master contact interface, col number is either the number of dofs of the master interior or contact interface